File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/inventory/EquipmentRecyclerViewAdapter.java
package com.habitrpg.android.habitica.ui.adapter.inventory;

import android.content.Context;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.models.inventory.Equipment;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.realm.OrderedRealmCollection;
import io.realm.RealmRecyclerViewAdapter;
import rx.Observable;
import rx.subjects.PublishSubject;

public class EquipmentRecyclerViewAdapter extends RealmRecyclerViewAdapter<Equipment, EquipmentRecyclerViewAdapter.GearViewHolder> {

    public String equippedGear;
    public Boolean isCostume;
    public String type;

    private PublishSubject<String> equipEvents = PublishSubject.create();

    public EquipmentRecyclerViewAdapter(@Nullable OrderedRealmCollection<Equipment> data, boolean autoUpdate) {
        super(data, autoUpdate);
    }

    @Override
    public GearViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.gear_list_item, parent, false);

        return new GearViewHolder(view);
    }

    @Override
    public void onBindViewHolder(GearViewHolder holder, int position) {
        if (getData() != null) {
            holder.bind(getData().get(position));
        }
    }

    public Observable<String> getEquipEvents() {
        return equipEvents.asObservable();
    }

    class GearViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.gear_container)
        View gearContainer;

        @BindView(R.id.gear_text)
        TextView gearNameTextView;

        @BindView(R.id.gear_notes)
        TextView gearNotesTextView;

        @BindView(R.id.gear_image)
        SimpleDraweeView imageView;

        @BindView(R.id.equippedIndicator)
        View equippedIndicator;

        Equipment gear;

        Context context;

        public GearViewHolder(View itemView) {
            super(itemView);

            ButterKnife.bind(this, itemView);

            context = itemView.getContext();
            itemView.setOnClickListener(this);
        }

        public void bind(Equipment gear) {
            this.gear = gear;
            this.gearNameTextView.setText(this.gear.text);
            this.gearNotesTextView.setText(this.gear.notes);

            if (gear.key.equals(equippedGear)) {
                this.equippedIndicator.setVisibility(View.VISIBLE);
                this.gearContainer.setBackgroundColor(ContextCompat.getColor(context, R.color.brand_700));
            } else {
                this.equippedIndicator.setVisibility(View.GONE);
                this.gearContainer.setBackgroundResource(R.drawable.selection_highlight);
            }

            String imageUrl = "https://habitica-assets.s3.amazonaws.com/mobileApp/images/shop_" + gear.key + ".png";
            imageView.setImageURI(Uri.parse(imageUrl));
        }

        @Override
        public void onClick(View v) {
            equipEvents.onNext(this.gear.key);
            if (this.gear.key.equals(equippedGear)) {
                equippedGear = type + "_base_0";
            } else {
                equippedGear = this.gear.key;
            }
            notifyDataSetChanged();
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/adapter/inventory/EquipmentRecyclerViewAdapter.kt
package com.habitrpg.android.habitica.ui.adapter.inventory

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.BitmapDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.databinding.GearListItemBinding
import com.habitrpg.android.habitica.models.inventory.Equipment
import com.habitrpg.android.habitica.ui.adapter.BaseRecyclerViewAdapter
import com.habitrpg.android.habitica.ui.views.HabiticaIconsHelper
import com.habitrpg.common.habitica.extensions.loadImage

class EquipmentRecyclerViewAdapter :
    BaseRecyclerViewAdapter<Equipment, EquipmentRecyclerViewAdapter.GearViewHolder>() {
    var equippedGear: String? = null
    var isCostume: Boolean? = null
    var type: String? = null

    var onEquip: ((String) -> Unit)? = null

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): GearViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.gear_list_item, parent, false)
        return GearViewHolder(view)
    }

    override fun onBindViewHolder(
        holder: GearViewHolder,
        position: Int
    ) {
        holder.bind(data[position])
    }

    inner class GearViewHolder(itemView: View) :
        androidx.recyclerview.widget.RecyclerView.ViewHolder(itemView) {
        private val binding = GearListItemBinding.bind(itemView)

        var gear: Equipment? = null
        var context: Context = itemView.context

        init {
            context = itemView.context
            binding.twoHandedView.setCompoundDrawablesWithIntrinsicBounds(
                BitmapDrawable(
                    context.resources,
                    HabiticaIconsHelper.imageOfTwoHandedIcon()
                ),
                null,
                null,
                null
            )
            itemView.setOnClickListener {
                val key = gear?.key
                if (key != null) {
                    onEquip?.invoke(key)
                    equippedGear =
                        if (key == equippedGear) {
                            type + "_base_0"
                        } else {
                            key
                        }
                    notifyDataSetChanged()
                }
            }
        }

        fun bind(gear: Equipment) {
            this.gear = gear
            binding.gearText.text = this.gear?.text
            binding.gearNotes.text = this.gear?.notes

            if (gear.key == equippedGear) {
                binding.equippedIndicator.visibility = View.VISIBLE
                binding.gearContainer.setBackgroundColor(
                    ContextCompat.getColor(
                        context,
                        R.color.lightly_tinted_background
                    )
                )
                binding.gearIconBackgroundView.background =
                    ContextCompat.getDrawable(context, R.drawable.layout_rounded_bg_content)
            } else {
                binding.equippedIndicator.visibility = View.GONE
                binding.gearContainer.setBackgroundResource(R.drawable.selection_highlight)
                binding.gearIconBackgroundView.background =
                    ContextCompat.getDrawable(context, R.drawable.layout_rounded_bg_window)
            }
            binding.twoHandedView.visibility = if (gear.twoHanded) View.VISIBLE else View.GONE
            binding.gearImage.loadImage("shop_" + gear.key)

            set(binding.strLabel, binding.strValue, gear.str)
            set(binding.conLabel, binding.conValue, gear.con)
            set(binding.intLabel, binding.intValue, gear.intelligence)
            set(binding.perLabel, binding.perValue, gear.per)
        }

        private fun set(
            label: TextView,
            valueTextView: TextView,
            value: Int
        ) {
            if (value > 0) {
                label.visibility = View.VISIBLE
                valueTextView.visibility = View.VISIBLE
                @SuppressLint("SetTextI18n")
                valueTextView.text = "+$value"
            } else {
                label.visibility = View.GONE
                valueTextView.visibility = View.GONE
            }
        }
    }
}
